### PR TITLE
Improve Llama2 and gpt_neox performance with Habana fused RoPE and RMSNorm

### PR DESCRIPTION
**Pull request by mandy-li**
# What does this PR do?

1. Improve Llama2 inference performance by using habana fused RoPE and RMSNorm
2. Improve gpt_neox inference performance by using habana fused RoPE

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -378,3 +385,10 @@ def prepare_inputs_for_generation(
             "past_key_values": past_key_values,
             "token_idx": token_idx,
         }
+
+
+def apply_customized_rope(q, k, cos, sin, position_ids):
+    if q.device.type == "hpu" and FusedRoPE:
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py:391]:
```suggestion
    if FusedRoPE:
```
q.device should always be 'hpu', means not support CPU or CUDA

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -238,6 +251,27 @@ def custom_forward(*inputs):
     )
 
 
+def gaudi_llama_rmsnorm_forward(self, hidden_states):
+    """
+    Copied from LlamaRMSNorm.forward: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py
+    The only differences are:
+        - override RMSNorm with Habana fused RMSNorm
+    """
+    if hidden_states.device.type == "hpu" and FusedRMSNorm:
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/llama/modeling_llama.py:260]:
```suggestion
    if FusedRMSNorm:
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -346,3 +380,10 @@ def prepare_inputs_for_generation(
             }
         )
         return model_inputs
+
+
+def apply_customized_rope(q, k, cos, sin, position_ids):
+    if q.device.type == "hpu" and FusedRoPE:
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/llama/modeling_llama.py:386]:
```suggestion
    if FusedRoPE:
```

File: optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 4]
```diff
@@ -6,6 +6,13 @@
 from transformers.models.gpt_neox.modeling_gpt_neox import GPTNeoXForCausalLM, apply_rotary_pos_emb, logger
 
 
+try:
+    from habana_frameworks.torch.hpex.kernels import RotaryPosEmbeddingHelperV2 as FusedRoPE
+except ImportError:
+    print("Not using HPU kernel for apply_rotary_pos_emb")
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py:12]:
```suggestion
    print("Not using HPU fused kernel for apply_rotary_pos_emb")
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -7,6 +7,19 @@
 from transformers.models.llama.modeling_llama import LlamaForCausalLM, apply_rotary_pos_emb, logger
 
 
+try:
+    from habana_frameworks.torch.hpex.kernels import RotaryPosEmbeddingHelperV2 as FusedRoPE
+except ImportError:
+    print("Not using HPU kernel for apply_rotary_pos_emb")
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/llama/modeling_llama.py:13]:
```suggestion
    print("Not using HPU fused kernel for apply_rotary_pos_emb")
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -7,6 +7,19 @@
...
+    FusedRoPE = None
+
+try:
+    from habana_frameworks.torch.hpex.normalization import FusedRMSNorm as FusedRMSNorm
+except ImportError:
+    print("Not using HPU kernel for RMSNorm")
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/llama/modeling_llama.py:19]:
```suggestion
    print("Not using HPU fused kernel for RMSNorm")
```

File: optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 1 above)
Comment (mandy-li) [reply optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py:391]: this optimum-habana script should be able to run on CPU, right?

**Comment by ZhaiFeiyue** [comment]
@mandy-li nice PR 👍

**Review by ZhaiFeiyue — APPROVED** [review]

File: optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 1 above)
Comment (mandy-li) [reply optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py:391]: Discussed with Feiyue offline, we decided to keep device type check in the source code.

**Review by regisss — APPROVED** [review]
LGTM!

**Pull request merged**